File: lyrimuse-collector/lyricsretryskip.go
package main

import "strings"

// 自动补空扫描与全量扫库要跳过的空歌词条目。手动「重试无歌词条目」不受影响 —— 用户点了就是要现在搜。
//
// 不按「数据缺失」一刀切:缺时长、缺专辑、连歌手都缺的条目照样常能搜到(靠歌名命中),本机缓存里
// 缺字段的条目有歌词的远多于没歌词的。跳过的只有两类「再搜也不会有」的:
//   - 没歌手、没专辑,补空已经失败 lyricsNoAnchorGiveUpCount 次:没有任何可依据的信息(播客单集、
//     电台节目之类),第一次靠歌名搜不到,之后也不会;
//   - 结构上是播放器把歌词写进身份留下的污染(见 lyricsPollutedKeys)。

// lyricsNoAnchorGiveUpCount:没歌手没专辑的空条目,补空失败几次后自动路径不再碰。
const lyricsNoAnchorGiveUpCount = 3

// lyricsPollutedMinVariants:同一个身份字段下另一个字段至少出现几个不同值,才算污染。
const lyricsPollutedMinVariants = 3

// lyricsNoAnchorGaveUp:没歌手、没专辑,补空已经失败够次数了。
func lyricsNoAnchorGaveUp(key string, e enrichEntry) bool {
	artist, _, album := splitEnrichKey(key)
	return strings.TrimSpace(artist) == "" && strings.TrimSpace(album) == "" &&
		e.LyricsFillCount >= lyricsNoAnchorGiveUpCount
}

// lyricsPollutedKeys 找出结构上是「一个字段装身份、另一个字段放歌词」留下的空歌词条目:同一张专辑下,
// 某个字段完全一样、而且能按带空格的分隔符拆成两段(塞着「歌名 - 歌手」),另一个字段却出现了
// lyricsPollutedMinVariants 个以上不同的值。信任进来的其他播放器出这种毛病时(见 trustedlyricartist.go
// 头注),纠正上线之前每句歌词都落了一条。
//
// 只看「同一歌手同一专辑下有很多不同歌名」不行,那就是一张正常专辑;所以要求不变的那个字段带分隔符
// —— 正常数据里歌手名几乎不带空格破折号,带破折号的歌名在同一张专辑下也只对应一个歌手。
// 调用方必须持有 enrichMu。
func lyricsPollutedKeys(cache map[string]enrichEntry) map[string]bool {
	type group struct {
		variants map[string]bool
		keys     []string
	}
	byTitle := map[string]*group{}
	byArtist := map[string]*group{}
	add := func(m map[string]*group, gk, variant, key string, empty bool) {
		g := m[gk]
		if g == nil {
			g = &group{variants: map[string]bool{}}
			m[gk] = g
		}
		g.variants[variant] = true
		if empty {
			g.keys = append(g.keys, key)
		}
	}
	for key, e := range cache {
		artist, title, album := splitEnrichKey(key)
		empty := e.Lyrics == "" && !e.ManualLyrics && !e.Instrumental
		if len(trustedSplitCandidates(title)) > 0 {
			add(byTitle, title+"|"+album, artist, key, empty)
		}
		if len(trustedSplitCandidates(artist)) > 0 {
			add(byArtist, artist+"|"+album, title, key, empty)
		}
	}
	out := map[string]bool{}
	for _, m := range []map[string]*group{byTitle, byArtist} {
		for _, g := range m {
			if len(g.variants) < lyricsPollutedMinVariants {
				continue
			}
			for _, k := range g.keys {
				out[k] = true
			}
		}
	}
	return out
}

File: lyrimuse/Sources/LyrimuseCore/Local/LyricsRetrySkip.swift
import Foundation

/// 全量重新扫库跳过的「再搜也不会有」的空歌词条目,collector 侧 `lyricsretryskip.go` 的镜像。
///
/// 界面「全量重新扫库」那一行的「N 首」要跟 collector 真会扫的条数逐条对得上(见 `LyricsFullScan.tier`),
/// 所以这两条判据两边必须同步改。手动「重试无歌词条目」不受它们影响,那个数不用它。
public enum LyricsRetrySkip {
    /// 没歌手没专辑的空条目,补空失败几次后自动路径不再碰。同 collector `lyricsNoAnchorGiveUpCount`。
    public static let noAnchorGiveUpCount = 3
    /// 同一个身份字段下另一个字段至少几个不同值才算污染。同 collector `lyricsPollutedMinVariants`。
    public static let pollutedMinVariants = 3

    /// 带空格的分隔符,同 collector `trustedTitleSeps`。
    static let separators = [" - ", " – ", " — ", " － "]

    /// 没歌手、没专辑,补空已经失败够次数了。
    public static func noAnchorGaveUp(artist: String, album: String, fillCount: Int) -> Bool {
        artist.trimmingCharacters(in: .whitespaces).isEmpty
            && album.trimmingCharacters(in: .whitespaces).isEmpty
            && fillCount >= noAnchorGiveUpCount
    }

    /// 能不能在某个带空格的分隔符处拆成两段、两段都有字母或数字 —— collector `trustedSplitCandidates`
    /// 非空的同一个判据。
    public static func hasSplit(_ s: String) -> Bool {
        for sep in separators {
            var searchStart = s.startIndex
            while let r = s.range(of: sep, range: searchStart..<s.endIndex) {
                if hasLetterOrDigit(s[s.startIndex..<r.lowerBound]) && hasLetterOrDigit(s[r.upperBound...]) {
                    return true
                }
                searchStart = s.index(after: r.lowerBound)
            }
        }
        return false
    }

    private static func hasLetterOrDigit(_ s: Substring) -> Bool {
        s.unicodeScalars.contains { CharacterSet.letters.contains($0) || CharacterSet.decimalDigits.contains($0) }
    }

    public struct Row: Sendable {
        public let key, artist, title, album: String
        /// 没歌词、没人工修正、没确证纯音乐。
        public let isEmpty: Bool
        public init(key: String, artist: String, title: String, album: String, isEmpty: Bool) {
            self.key = key
            self.artist = artist
            self.title = title
            self.album = album
            self.isEmpty = isEmpty
        }
    }

    /// 结构上是「一个字段装身份、另一个字段放歌词」留下的空歌词条目,判据同 collector `lyricsPollutedKeys`:
    /// 同一张专辑下,某个字段完全一样且能拆成两段,另一个字段出现 `pollutedMinVariants` 个以上不同值。
    public static func pollutedKeys(_ rows: [Row]) -> Set<String> {
        var byTitle: [String: (variants: Set<String>, keys: [String])] = [:]
        var byArtist: [String: (variants: Set<String>, keys: [String])] = [:]
        for row in rows {
            if hasSplit(row.title) {
                let gk = row.title + "|" + row.album
                byTitle[gk, default: ([], [])].variants.insert(row.artist)
                if row.isEmpty { byTitle[gk, default: ([], [])].keys.append(row.key) }
            }
            if hasSplit(row.artist) {
                let gk = row.artist + "|" + row.album
                byArtist[gk, default: ([], [])].variants.insert(row.title)
                if row.isEmpty { byArtist[gk, default: ([], [])].keys.append(row.key) }
            }
        }
        var out = Set<String>()
        for group in Array(byTitle.values) + Array(byArtist.values) where group.variants.count >= pollutedMinVariants {
            out.formUnion(group.keys)
        }
        return out
    }
}
